### PR TITLE
Improve vstudio

### PR DESCRIPTION
**Pull request by SirLynix**
Improve vstudio projects:
- setup multi-core compilation by default (no longer require `add_cxflags("/MP")`, which is VS default behavior) - can be disabled with flag /MP1
- fixes external include dirs, which were not translated like regular include dirs are.
- add support for rundir
- add support for /W4 ("allextra" warning level)

Also during my tests I noticed that some flags may appear multiple times, and that the common flag check (which is also done in makefile project) failed because some flags appeared multiple times (like `-experimental:external` and such, which I suppose are added after xmake remove duplicates flags).

Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags

### CODE REVIEW

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 1]
```diff
@@ -528,6 +550,7 @@ function _make_common_items(vcxprojfile, vsinfo, target, vcxprojdir)
 
                     -- no common flags for asm
                     if sourcekind ~= "as" then
+                        local foundDebug = false
```
Comment (waruqi) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:553]: unused var? and please use lowercase

**Comment by waruqi** [comment]
> Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags

it will remove repeat flags.

https://github.com/xmake-io/xmake/blob/f3b110572f3720905fc3ee97362d6055f5f29d78/xmake/core/tool/builder.lua#L358

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua (hunk 1 above)
Comment (SirLynix) [reply xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:553]: I removed it

**Comment by SirLynix** [comment]
> > Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags
> 
> it will remove duplicate flags.
> 
> https://github.com/xmake-io/xmake/blob/f3b110572f3720905fc3ee97362d6055f5f29d78/xmake/core/tool/builder.lua#L358

It doesn't because external headers are passed as a table, and thus concat. So it doesn't see "-external:W0" but "-external:W0-external:I..."

**Comment by waruqi** [comment]
> > > Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags
> > 
> > 
> > it will remove duplicate flags.
> > https://github.com/xmake-io/xmake/blob/f3b110572f3720905fc3ee97362d6055f5f29d78/xmake/core/tool/builder.lua#L358
> 
> It doesn't because external headers are passed as a table, and thus concat. So it doesn't see "-external:W0" but "-external:W0-external:I..."

The table is used to avoid internal deduplication, for example: avoid `-I xx -I yy` -> `-I xx yy`, there is no problem, unless "-external:W0" and "-external:-I" are not put in a table.

**Comment by SirLynix** [comment]
> > > > Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags
> > > 
> > > 
> > > it will remove duplicate flags.
> > > https://github.com/xmake-io/xmake/blob/f3b110572f3720905fc3ee97362d6055f5f29d78/xmake/core/tool/builder.lua#L358
> > 
> > 
> > It doesn't because external headers are passed as a table, and thus concat. So it doesn't see "-external:W0" but "-external:W0-external:I..."
> 
> The table is used to avoid internal deduplication, for example: avoid `-I xx -I yy` -> `-I xx yy`, there is no problem, unless "-external:W0" and "-external:-I" are not put in a table.

I understand what purpose it serves, but it will duplicate -external:W0, see this:

```
[ 32%]: compiling.debug src\Nazara\Utility\Formats\FreeTypeLoader.cpp
"C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.30.30705\\bin\\HostX64\\x64\\cl.exe" -c /EHsc -nologo -Zi -FS -Fdbin\windows_x64_debug\compile.NazaraUtility-d.pdb -W4 -Od -std:c++17 -MDd -Iinclude -Isrc -D_CRT_SECURE_NO_WARNINGS -DNAZARA_BUILD -DNAZARA_UTILITY_BUILD -DNAZARA_DEBUG -DNAZARA_UTILITY_DEBUG -DBZ_STATIC -external:W0 -external:Ithirdparty\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\e\entt\v3.8.1\b21698e7653b48c49e073d6614de13b4\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\f\freetype\2.11.0\ee2054a878394e47ba6e975a91854c2b\include\freetype2 -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\b\brotli\1.0.9\b496cb273fdb41089073359cc81ebb58\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\l\libpng\v1.6.37\d3e3dccbfcec47738b9bb4a32a0a1f44\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\z\zlib\1.2.11\d6b178abc5d04bdc936c37166df8fe84\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\b\bzip2\1.0.8\cc4412cb62de4ce68f2fe59e6d7f8704\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\s\stb\0.0\86911977ee9f47b7ac1685c567fd7b8b\include /bigobj /permissive- /Zc:__cplusplus /Zc:externConstexpr /Zc:inline /Zc:lambda /Zc:preprocessor /Zc:referenceBinding /Zc:strictStrings /Zc:throwingNew /w44062 /wd4251 -Fobuild\.objs\NazaraUtility\windows\x64\debug\src\Nazara\Utility\Formats\FreeTypeLoader.cpp.obj src\Nazara\Utility\Formats\FreeTypeLoader.cpp
```

`-external:W0` is put before every `-external:I` because they're in a table and thus not deduplicated.

**Comment by waruqi** [comment]
> > > > > Edit: I checked the reason why some duplicate flags are still there and it's because they were added as a parameter table, thus handled as a whole by _preprocess_flags
> > > > 
> > > > 
> > > > it will remove duplicate flags.
> > > > https://github.com/xmake-io/xmake/blob/f3b110572f3720905fc3ee97362d6055f5f29d78/xmake/core/tool/builder.lua#L358
> > > 
> > > 
> > > It doesn't because external headers are passed as a table, and thus concat. So it doesn't see "-external:W0" but "-external:W0-external:I..."
> > 
> > 
> > The table is used to avoid internal deduplication, for example: avoid `-I xx -I yy` -> `-I xx yy`, there is no problem, unless "-external:W0" and "-external:-I" are not put in a table.
> 
> I understand what purpose it serves, but it will duplicate -external:W0, see this:
> 
> ```
> [ 32%]: compiling.debug src\Nazara\Utility\Formats\FreeTypeLoader.cpp
> "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.30.30705\\bin\\HostX64\\x64\\cl.exe" -c /EHsc -nologo -Zi -FS -Fdbin\windows_x64_debug\compile.NazaraUtility-d.pdb -W4 -Od -std:c++17 -MDd -Iinclude -Isrc -D_CRT_SECURE_NO_WARNINGS -DNAZARA_BUILD -DNAZARA_UTILITY_BUILD -DNAZARA_DEBUG -DNAZARA_UTILITY_DEBUG -DBZ_STATIC -external:W0 -external:Ithirdparty\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\e\entt\v3.8.1\b21698e7653b48c49e073d6614de13b4\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\f\freetype\2.11.0\ee2054a878394e47ba6e975a91854c2b\include\freetype2 -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\b\brotli\1.0.9\b496cb273fdb41089073359cc81ebb58\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\l\libpng\v1.6.37\d3e3dccbfcec47738b9bb4a32a0a1f44\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\z\zlib\1.2.11\d6b178abc5d04bdc936c37166df8fe84\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\b\bzip2\1.0.8\cc4412cb62de4ce68f2fe59e6d7f8704\include -external:W0 -external:IC:\Users\Lynix\AppData\Local\.xmake\packages\s\stb\0.0\86911977ee9f47b7ac1685c567fd7b8b\include /bigobj /permissive- /Zc:__cplusplus /Zc:externConstexpr /Zc:inline /Zc:lambda /Zc:preprocessor /Zc:referenceBinding /Zc:strictStrings /Zc:throwingNew /w44062 /wd4251 -Fobuild\.objs\NazaraUtility\windows\x64\debug\src\Nazara\Utility\Formats\FreeTypeLoader.cpp.obj src\Nazara\Utility\Formats\FreeTypeLoader.cpp
> ```
> 
> `-external:W0` is put before every `-external:I` because they're in a table and thus not deduplicated.

I know it will be repeated, but there is no better way, they can only be placed in a table. https://github.com/xmake-io/xmake/blob/5c4adb14e1f45b40bb3d90a9459386f67f08c15f/xmake/modules/core/tools/cl.lua#L284

**Comment by SirLynix** [comment]
> I know it will be repeated, but there is no better way, they can only be placed in a table.
> 
> https://github.com/xmake-io/xmake/blob/5c4adb14e1f45b40bb3d90a9459386f67f08c15f/xmake/modules/core/tools/cl.lua#L284

Yes, but then it will break makefile and vstudio common flags search, because they will have a flag more than they have files, this is why I fixed it using a `>=` instead of `==`

**Comment by waruqi** [comment]
> Yes, but then it will break makefile and vstudio common flags search, because they will have a flag more than they have files, this is why I fixed it using a `>=` instead of `==`

You can use this workaround solution to be compatible with it first

**Pull request merged**